tachyon/crypto/commitments/fri/fri.h

```diff
@@ -0,0 +1,221 @@
+// Use of this source code is governed by a Apache-2.0 style license that
+// can be found in the LICENSE.lambdaworks.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_FRI_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_FRI_H_
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/crypto/commitments/fri/fri_proof.h"
+#include "tachyon/crypto/commitments/fri/fri_storage.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree.h"
+#include "tachyon/crypto/commitments/univariate_polynomial_commitment_scheme.h"
+#include "tachyon/crypto/transcripts/transcript.h"
+
+namespace tachyon::crypto {
+
+template <typename F, size_t MaxDegree>
+class FRI : public UnivariatePolynomialCommitmentScheme<FRI<F, MaxDegree>> {
+ public:
+  using Base = UnivariatePolynomialCommitmentScheme<FRI<F, MaxDegree>>;
+  using Poly = typename Base::Poly;
+  using Evals = typename Base::Evals;
+  using Domain = typename Base::Domain;
+
+  FRI() = default;
+  FRI(const Domain* domain, FRIStorage<F>* storage,
+      BinaryMerkleHasher<F, F>* hasher)
+      : domain_(domain),
+        storage_(storage),
+        hasher_(hasher),
+        two_inv_(F(2).Inverse()) {
+    CHECK_GE(domain->size(), size_t{2}) << "This ensures last folding process";
+    size_t k = domain->log_size_of_group();
+    if (k > 1) {
+      sub_domains_ = base::CreateVector(k - 1, [k](size_t i) {
+        return Domain::Create(size_t{1} << (k - i - 1));
+      });
+    }
+    storage_->Allocate(k);
+  }
+
+  // UnivariatePolynomialCommitmentScheme methods
```

Comment:
This is a method of the `FRI`, and it appears that the base classes are using `FRI::N()`. Please confirm if the meaning of this comment is correct.

Reply:
I follow this convention when I write some methods which are either or virtual methods or needed by parent class